sound/soc/sof/intel/hda.c

```diff
@@ -470,10 +475,28 @@ static int hda_init_caps(struct snd_sof_dev *sdev)
 	else
 		dev_info(bus->dev, "hda codecs found, mask %lx!\n", bus->codec_mask);
 
-	/* used by hda machine driver to create dai links */
-	mach_params = (struct snd_soc_acpi_mach_params *)
-		&sdev->pdata->machine->mach_params;
-	mach_params->codec_mask = bus->codec_mask;
+	if (bus->codec_mask) {
+		for (i = 0; i < 64; i++) {
+			if (bus->codec_mask & (1 << i))
+				codec_num++;
+		}
+
+		/* if there are HDMI codec and HDA codecs, hda machine
+		 * is applied. For I2S codec + HDMI codec, there is only one
+		 * hda codec. For other cases, the value should be zero.
+		 */
+		if (codec_num > 1) {
+			/* used by hda machine driver to create dai links */
+			mach = snd_soc_acpi_intel_hda_machines;
```

Comment:
@RanderWang do we need to care about freeing the memory allocated to mach before doing this?

Reply:
yes, good point